### PR TITLE
Add `ElemNetFeaturizer`

### PR DESCRIPTION
**Pull request by aksub99**
Add featurizer to convert pymatgen composition into vector of fractional compositions (As utilized in https://www.nature.com/articles/s41598-018-35934-y)

TODO:
- [x] Fix type annotations
- [x] Convert `data_utils` code into `Pymatgen` API.
- [x] Add unit tests

### CODE REVIEW

File: deepchem/feat/material_featurizers/composition_featurizer.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+  Examples
+  --------
+  >>> import pymatgen as mg
+  >>> comp = mg.Composition("Fe2O3")
+  >>> featurizer = CompositionFeaturizer()
+  >>> features = featurizer.featurize([comp])
+  """
+
+  def get_fractions(self, comp):
```
Comment (nd-02110114) [inline deepchem/feat/material_featurizers/composition_featurizer.py:43]: Can you add docstring and type annotation?

File: deepchem/feat/material_featurizers/composition_featurizer.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    if all(e in elements_tl for e in comp):
+      return np.array([comp[e] if e in comp else 0 for e in elements_tl],
+                      np.float32)
+    else:
+      return None
+
+  def parse_fractions(self, form):
```
Comment (nd-02110114) [inline deepchem/feat/material_featurizers/composition_featurizer.py:50]: Can you add docstring and type annotation?

File: deepchem/feat/material_featurizers/composition_featurizer.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    while '/' in form:
+      di = form.index('/')
+      num1 = [x for x in re.findall(r'\d*\.*\d*', form[:di]) if x != ''][-1]
+      num2 = [x for x in re.findall(r'\d*\.*\d*', form[di + 1:]) if x != ''][0]
+      fract = '%.3f' % (float(num1) / float(num2))
+      form = form[:di - len(num1)] + fract + form[di + len(num2) + 1:]
+    return form
+
+  def parse_formula(self, formula):
```
Comment (nd-02110114) [inline deepchem/feat/material_featurizers/composition_featurizer.py:59]: Same here.

File: deepchem/feat/material_featurizers/composition_featurizer.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+
+class CompositionFeaturizer(MaterialCompositionFeaturizer):
+  """
+  Fixed size vector containing raw elemental compositions in the compound.
+
+  Returns a vector containing fractional compositions of each element
+  in the compound.
+
+  This featurizer requires the optional dependency pymatgen.
```
Comment (nd-02110114) [inline deepchem/feat/material_featurizers/composition_featurizer.py:28]: Can you change like this?
https://github.com/deepchem/deepchem/blob/ab07c0b6b070d376039f55762e6894df289b7364/deepchem/feat/material_featurizers/element_property_fingerprint.py#L40-L42

File: deepchem/feat/material_featurizers/composition_featurizer.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    Returns
+    -------
+    feats: np.ndarray
+      Vector of fractional compositions of each element.
+    """
+    try:
+      pretty_comp = composition.reduced_formula
+      feats = self.get_fractions(self.parse_formula(pretty_comp))
+    except:
```
Comment (nd-02110114) [inline deepchem/feat/material_featurizers/composition_featurizer.py:149]: We don't need to catch the exception. `MaterialCompositionFeaturizer` (parent class) catch the exception.

**Review by nd-02110114 — COMMENTED** [review]
Thank you for contribution! 
I seem this implementation was based on https://github.com/NU-CUCIS/ElemNet/blob/master/elemnet/data_utils.py. 
Can you add more explanation about the ElemNet and the feature vector this class calculates?

File: deepchem/feat/material_featurizers/composition_featurizer.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    'Ru', 'Rh', 'Pd', 'Ag', 'Cd', 'In', 'Sn', 'Sb', 'Te', 'I', 'Xe', 'Cs', 'Ba',
+    'La', 'Ce', 'Pr', 'Nd', 'Pm', 'Sm', 'Eu', 'Gd', 'Tb', 'Dy', 'Ho', 'Er',
+    'Tm', 'Yb', 'Lu', 'Hf', 'Ta', 'W', 'Re', 'Os', 'Ir', 'Pt', 'Au', 'Hg', 'Tl',
+    'Pb', 'Bi', 'Ac', 'Th', 'Pa', 'U', 'Np', 'Pu'
+]
+
+formulare = re.compile(r'([A-Z][a-z]*)(\d*\.*\d*)')
+
+
+class CompositionFeaturizer(MaterialCompositionFeaturizer):
```
Comment (rbharath) [inline deepchem/feat/material_featurizers/composition_featurizer.py:21]: Could you add unit tests for `CompositionFeaturizer`?

File: deepchem/feat/material_featurizers/__init__.py [hunk 7]
```diff
@@ -5,3 +5,4 @@
 from deepchem.feat.material_featurizers.element_property_fingerprint import ElementPropertyFingerprint
 from deepchem.feat.material_featurizers.sine_coulomb_matrix import SineCoulombMatrix
 from deepchem.feat.material_featurizers.cgcnn_featurizer import CGCNNFeaturizer
+from deepchem.feat.material_featurizers.composition_featurizer import CompositionFeaturizer
```
Comment (rbharath) [inline deepchem/feat/material_featurizers/__init__.py:8]: Could you add `CompositionFeaturizer` to the featurization api docs in `docs/source/api_reference/featurizers.rst`?

**Review by rbharath — COMMENTED** [review]
Thanks for the contribution! Have a couple of requests for tests/docs as well

**Comment by aksub99** [comment]
Thanks for the reviews @nd-02110114 and @rbharath ! 
I'm done with most of the modifications. Only unit tests are pending.

**Comment by nd-02110114** [comment]
I confirmed [data_preprocess.ipynb](https://github.com/NU-CUCIS/ElemNet/blob/master/elemnet/data_preprocess.ipynb) and other some files in ElemNet repo and I came up with some questions.


1. What is your usage of this featurizer? Your implementation, extracting the fraction of each element from the given composition, is a  part of preprocess steps  on ElemNet. This means the feature this featurizer returns is not used as the input of models. Basically, the featurizer returns a feature which is used as an input.

2. I think we get the fraction of each element easily from pymatgen API. See, https://pymatgen.org/pymatgen.core.composition.html?highlight=fractional#pymatgen.core.composition.Composition.fractional_composition. The authors use your implementations to parse various dirty compositions like '(3InAs)0.95(In2Te3)0.05', 'Mg1/3Ta2/3' which pymatgen can't parse. However, your featurizer receive the pymatgen composition object. Pymatgen have already parsed the raw composition in the parent class, `MaterialCompositionFeaturizer`. So, I think most of your implementation is not needed.

I will restart reviewing after this question is resolved.

**Comment by aksub99** [comment]
Thanks for your questions @nd-02110114 !
1. Yes, this code is used as a part of the data preprocessing steps. The output of the featurizer would be a part of the input to the model, it wouldn't define the complete input to the model though. Other features would have to be concatenated with this vector in order to make it usable by the model. 
My plan was to add the `oqmd_all-22Mar18.csv` file (used as the initial dataset in `data_preprocess.ipynb`) to MoleculeNet and use the defined featurizer to featurize the composition. I was thinking that the other preprocessing steps could be done by using appropriate transformers. Please let me know if you think an alternative direction would be a better option.
In fact, I saw that the preprocessed train and test sets are also available. Would directly adding that to MoleculeNet be a better option?

2. `data_preprocess.ipynb` also starts with the pymatgen composition object and then uses the code to handle dirty compositions like the ones you have mentioned. Since they start with the composition object as well, I assumed that pymatgen's `fractional_composition` function didn't take care of dirty compositions under the hood. So, I felt that this code was necessary. Am I misunderstanding your point?

**Comment by nd-02110114** [comment]
>  I was thinking that the other preprocessing steps could be done by using appropriate transformers.

I think it is better to include all preprocessing steps in this featurizer. This is because basically all featuriezers return the feature as the model's input. 

> In fact, I saw that the preprocessed train and test sets are also available. Would directly adding that to MoleculeNet be a better option?

I think it is OK. But, it is more useful to implement the featurizer which contains all preprocessing steps.

> data_preprocess.ipynb also starts with the pymatgen composition object and then uses the code to handle dirty compositions like the ones you have mentioned. Since they start with the composition object as well, I assumed that pymatgen's fractional_composition function didn't take care of dirty compositions under the hood. So, I felt that this code was necessary. Am I misunderstanding your point?

In data_preprocess.ipynb, the authors use the utilities which handle dirty compositions like `data_utils. parse_formula`.  But, the usage is not for handling dirty compositions. They were used for getting just the fractional composition.  Please confirm compositions in `oqmd_all-22Mar18.csv`. All compositions are not dirty, and they can be parsed by Pymatgen. And, I created the notebook which shows Pymatgen `fractional_composition` API returns the same value which `data_utils.pasre_formula` returns.

https://colab.research.google.com/drive/1uG7tFLGXeqeUOQzednnVAPgMdy3cAJu8?usp=sharing

**Comment by aksub99** [comment]
Ah, I see. That makes sense. I'll add in all the preprocessing steps to this featurizer.

Oh, good catch! I didn't realize that there weren't any dirty compositions in `oqmd_all-22Mar18.csv`. `Pymatgen`'s API definitely seems to be the better choice then. Thanks for the pointers.

**Comment by aksub99** [comment]
@nd-02110114 Quick question: There are preprocessing steps that make use of other columns in the csv file (such as removing outliers based on the `delta_e` column). 
From what I understand, featurizers seem to take only structures/graphs/compositions as input. So, how can I proceed with these other preprocessing steps that require access to non-structure columns?

**Comment by nd-02110114** [comment]
@aksub99 
I'm so sorry for a late reply 🙇‍♂️ . I misunderstood the preprocessing steps. You don't need to implement preprocessing steps that make use of other columns in the csv file. I think this featurizer only perform extracting the fraction of each element. 

So, if you add the test, this PR is ready to merge! And can you change the name of the featurizer from  `CompositionFeaturizer` to `ElemNetFeaturizer`? I seem that the name of `CompositionFeaturizer` is a little ambiguous.

**Comment by aksub99** [comment]
@nd-02110114 Thanks for the reply.
Sure, sounds good!
Makes sense, will modify the name accordingly.

**Comment by aksub99** [comment]
Looks as though the travis failures are unrelated to this PR.

**Comment by nd-02110114** [comment]
Thank you for all fix!
I think this PR is ready to merge!

And, if your next action is adding `oqmd_all-22Mar18.csv` to the MoleculeNet, I think [this file](https://github.com/deepchem/deepchem/blob/master/deepchem/molnet/load_function/bace_datasets.py) may be helpful for you.
Currently, we reimplement MoleculeNet loader functions.  All files in [molnet/load_function/material_datasets](https://github.com/deepchem/deepchem/tree/master/deepchem/molnet/load_function/material_datasets) is old style implementations. Please don't use the old style.

**Comment by rbharath** [comment]
Looks like the travis failures are caused by flake8 and flaky tests unrelated to this PR so I think we're good to go ahead and merge in. Congrats on the merged PR! :)

**Pull request merged**